File: heaps/javascript/heap.js
class MaxHeap {
    constructor(items = []) {
        this.heap = [];

        if (items.length > 0) {
            this.buildHeap(items);
        }
    }

    parent(index) {
        return Math.floor((index - 1) / 2);
    }

    left(index) {
        return 2 * index + 1;
    }

    right(index) {
        return 2 * index + 2;
    }

    swap(i, j) {
        [this.heap[i], this.heap[j]] = [this.heap[j], this.heap[i]];
    }

    heapify(index) {
        let largest = index;

        const left = this.left(index);
        const right = this.right(index);

        if (
            left < this.heap.length &&
            this.heap[left] > this.heap[largest]
        ) {
            largest = left;
        }

        if (
            right < this.heap.length &&
            this.heap[right] > this.heap[largest]
        ) {
            largest = right;
        }

        if (largest !== index) {
            this.swap(index, largest);
            this.heapify(largest);
        }
    }

    buildHeap(items) {
        this.heap = [...items];

        const start = Math.floor(this.heap.length / 2) - 1;

        for (let i = start; i >= 0; i--) {
            this.heapify(i);
        }
    }

    peek() {
        if (this.heap.length === 0) {
            return null;
        }

        return this.heap[0];
    }

    insert(value) {
        this.heap.push(value);

        let index = this.heap.length - 1;

        while (index > 0) {
            const parent = this.parent(index);

            if (this.heap[parent] < this.heap[index]) {
                this.swap(parent, index);
                index = parent;
            } else {
                break;
            }
        }
    }

    extractMax() {
        if (this.heap.length === 0) {
            return null;
        }

        if (this.heap.length === 1) {
            return this.heap.pop();
        }

        const maxValue = this.heap[0];

        this.heap[0] = this.heap.pop();

        this.heapify(0);

        return maxValue;
    }

    extractMaxN(n) {
        const result = [];

        const count = Math.min(n, this.heap.length);

        for (let i = 0; i < count; i++) {
            result.push(this.extractMax());
        }

        return result;
    }

    search(value) {
        for (let i = 0; i < this.heap.length; i++) {
            if (this.heap[i] === value) {
                return i;
            }
        }

        return -1;
    }

    delete(value) {
        const index = this.search(value);

        if (index === -1) {
            return false;
        }

        const lastIndex = this.heap.length - 1;

        if (index === lastIndex) {
            this.heap.pop();
            return true;
        }

        this.heap[index] = this.heap.pop();

        const parent = this.parent(index);

        if (
            index > 0 &&
            this.heap[index] > this.heap[parent]
        ) {
            let current = index;

            while (current > 0) {
                const parentIndex = this.parent(current);

                if (
                    this.heap[parentIndex] <
                    this.heap[current]
                ) {
                    this.swap(parentIndex, current);
                    current = parentIndex;
                } else {
                    break;
                }
            }
        } else {
            this.heapify(index);
        }

        return true;
    }

    size() {
        return this.heap.length;
    }

    isEmpty() {
        return this.heap.length === 0;
    }

    printHeap() {
        console.log(this.heap);
    }
}


const heap = new MaxHeap([3, 10, 5, 6, 2, 8, 1]);

console.log("Initial Heap:");
heap.printHeap();

console.log("\nPeek:");
console.log(heap.peek());

console.log("\nInsert 15:");
heap.insert(15);
heap.printHeap();

console.log("\nExtract Max:");
console.log(heap.extractMax());
heap.printHeap();

console.log("\nExtract Top 3:");
console.log(heap.extractMaxN(3));
heap.printHeap();

console.log("\nSearch 5:");
console.log(heap.search(5));

console.log("\nDelete 5:");
heap.delete(5);
heap.printHeap();

File: heaps/python/heap.py
class MaxHeap:
    def __init__(self, items=None):
        self.heap = []
        if items:
            self.build_heap(items)

    def parent(self, index):
        return (index - 1) // 2

    def left(self, index):
        return 2 * index + 1

    def right(self, index):
        return 2 * index + 2

    def swap(self, i, j):
        self.heap[i], self.heap[j] = self.heap[j], self.heap[i]

    def heapify(self, index):
        size = len(self.heap)
        largest = index

        left = self.left(index)
        right = self.right(index)

        if left < size and self.heap[left] > self.heap[largest]:
            largest = left

        if right < size and self.heap[right] > self.heap[largest]:
            largest = right

        if largest != index:
            self.swap(index, largest)
            self.heapify(largest)

    def build_heap(self, items):
        self.heap = list(items)

        start = (len(self.heap) // 2) - 1

        for i in range(start, -1, -1):
            self.heapify(i)

    def peek(self):
        if not self.heap:
            return None
        return self.heap[0]

    def insert(self, value):
        self.heap.append(value)
        index = len(self.heap) - 1

        while index > 0:
            parent = self.parent(index)

            if self.heap[parent] < self.heap[index]:
                self.swap(parent, index)
                index = parent
            else:
                break

    def extract_max(self):
        if not self.heap:
            return None

        if len(self.heap) == 1:
            return self.heap.pop()

        max_value = self.heap[0]

        self.heap[0] = self.heap.pop()

        self.heapify(0)

        return max_value

    def extract_max_n(self, n):
        result = []

        for _ in range(min(n, len(self.heap))):
            result.append(self.extract_max())

        return result

    def search(self, value):
        for index, item in enumerate(self.heap):
            if item == value:
                return index
        return -1

    def delete(self, value):
        index = self.search(value)

        if index == -1:
            return False

        last_index = len(self.heap) - 1

        if index == last_index:
            self.heap.pop()
            return True

        self.heap[index] = self.heap.pop()

        parent = self.parent(index)

        if index > 0 and self.heap[index] > self.heap[parent]:
            while index > 0:
                parent = self.parent(index)

                if self.heap[parent] < self.heap[index]:
                    self.swap(parent, index)
                    index = parent
                else:
                    break
        else:
            self.heapify(index)

        return True

    def size(self):
        return len(self.heap)

    def is_empty(self):
        return len(self.heap) == 0

    def print_heap(self):
        print(self.heap)


heap = MaxHeap([3, 10, 5, 6, 2, 8, 1])

print("Initial Heap:")
heap.print_heap()

print("\nPeek:")
print(heap.peek())

print("\nInsert 15:")
heap.insert(15)
heap.print_heap()

print("\nExtract Max:")
print(heap.extract_max())
heap.print_heap()

print("\nExtract Top 3:")
print(heap.extract_max_n(3))
heap.print_heap()

print("\nSearch 5:")
print(heap.search(5))

print("\nDelete 5:")
heap.delete(5)
heap.print_heap()

